Replace the use of std::list with std::vector. This avoids a lof of n…

baldersheim opened the pull request:
…ew/delete.

The number of elements that are moved around is small and significantly less than
the gain by more efficient memory management.

@vekterli PR This brought down execution time from 400s to 120s.

baldersheim committed:
Replace the use of std::list with std::vector. This avoids a lof of new/delete.
The number of elements that are moved around is small and significantly less than
the gain by more efficient memory management.

vekterli commented on vdslib/src/vespa/vdslib/distribution/distribution.cpp:
```diff
@@ -443,7 +429,7 @@ Distribution::getIdealNodes(const NodeType& nodeType,
         // for cheap access to back(). Stuff in redundancy fake entries to
         // avoid needing to check size during iteration.
-        std::list<ScoredNode> tmpResults(groupRedundancy, ScoredNode(0, 0, 0));
+        std::vector<ScoredNode> tmpResults(groupRedundancy, ScoredNode(0, 0, 0));
```
Consider updating comment that explicitly mentions linked list usage

baldersheim replied:
Comment updated, and some other related improvements.

vekterli reviewed (approved):
:+1:

baldersheim committed:
Allocate once and reuse.

baldersheim commented:
@vekterli PTAL

vekterli commented on vdslib/src/vespa/vdslib/distribution/distribution.cpp:
```diff
@@ -459,12 +460,12 @@ Distribution::getIdealNodes(const NodeType& nodeType,
                 }
                 if (it == tmpResults.end()) {
-                    tmpResults.emplace_back(nodes[j], nodeState.getReliability(), score);
+                    tmpResults.emplace_back(score, nodes[j], nodeState.getReliability());
```
Hmm is it technically possible for `it` to be invalidated by the `insert` above, i.e. comparing a bogus iterator with `end`? It _probably_ won't be since it's pre-reserved, but I'm not sure what guarantees are given here... Maybe check on something else than the iterator to be sure? If this is pulled out into its own function then could return instead of break and avoid the check entirely.

baldersheim replied:
Fixed

vekterli reviewed (commented):
A potential edge case to be pondered first, otherwise LGTM

baldersheim committed:
- Factor out insert() to a separate method.
- Inline Node::Node.
- Remove virtuality of Node
- Inline State::oneOf
- Microoptimize ClusterState::getNodeState.
This brings the runtime down from 2.23s to 1.68s And with valgrind from 130s to 85s.
In total with the other commits in this PR it is now down from 2.95s to 1.68, and from 360s to 85s with valgrind.

baldersheim commented:
@vekterli PTAL I found loads of other inefficient code too.

vekterli commented on vdslib/src/vespa/vdslib/state/clusterstate.cpp:
```diff
@@ -280,34 +280,40 @@ ClusterState::getNodeCount(const NodeType& type) const
     if (it != _nodeStates.end()) return it->second;
+
+    // If beyond node count, the node is down.
+    if (__builtin_expect(node.getIndex() >= _nodeCount[node.getType()], false)) {
```
Is it cheaper in practice to always do the map lookup than to have this check prior to it?

baldersheim replied:
It layed out the code in a more efficient way. There seemed to be less jumps back and forth.
The most important change was moving the throwing out to separate non inlined method as it eliminated the need for pushing and popping on the stack.
The second one was reorganising to only call throwUnknownType in one place.
And then I assumed it is more likely the node is in the _nodeStates map, than being beyond the node count.

vekterli reviewed (approved):
:+1:
